inference-engine/src/mkldnn_plugin/nodes/non_max_suppression.cpp

```diff
@@ -135,110 +213,202 @@ class NonMaxSuppressionImpl: public ExtLayerBase {
+                              inputs[NMS_SOFTNMSSIGMA]->getTensorDesc().getBlockingDesc().getOffsetPadding())[0];
+        float scale = 0.0f;
+        if (soft_nms_sigma > 0.0) {
+            scale = -0.5 / soft_nms_sigma;
+        }
+
+        int *selected_indices = outputs[NMS_SELECTEDINDICES]->buffer().as<int *>() +
+                                outputs[NMS_SELECTEDINDICES]->getTensorDesc().getBlockingDesc().getOffsetPadding();
+
+        float *selected_scores = nullptr;
+        if (outputs.size() > 1)
+            selected_scores = outputs[NMS_SELECTEDSCORES]->buffer().as<float *>() +
+                              outputs[NMS_SELECTEDSCORES]->getTensorDesc().getBlockingDesc().getOffsetPadding();
 
-        SizeVector boxesStrides = inputs[NMS_BOXES]->getTensorDesc().getBlockingDesc().getStrides();
-        SizeVector scoresStrides = inputs[NMS_SCORES]->getTensorDesc().getBlockingDesc().getStrides();
+        int *valid_outputs = nullptr;
+        if (outputs.size() > 2)
+            valid_outputs = outputs[NMS_VALIDOUTPUTS]->buffer().as<int *>() +
+                            outputs[NMS_VALIDOUTPUTS]->getTensorDesc().getBlockingDesc().getOffsetPadding();
+
+        const SizeVector &boxesStrides = inputs[NMS_BOXES]->getTensorDesc().getBlockingDesc().getStrides();
+        const SizeVector &scoresStrides = inputs[NMS_SCORES]->getTensorDesc().getBlockingDesc().getStrides();
 
         // boxes shape: {num_batches, num_boxes, 4}
         // scores shape: {num_batches, num_classes, num_boxes}
-        int num_batches = static_cast<int>(scores_dims[0]);
-        int num_classes = static_cast<int>(scores_dims[1]);
-        std::vector<filteredBoxes> fb;
-
-        for (int batch = 0; batch < num_batches; batch++) {
-            float *boxesPtr = boxes + batch * boxesStrides[0];
-            for (int class_idx = 0; class_idx < num_classes; class_idx++) {
-                float *scoresPtr = scores + batch * scoresStrides[0] + class_idx * scoresStrides[1];
-                std::vector<std::pair<float, int> > scores_vector;
-                for (int box_idx = 0; box_idx < num_boxes; box_idx++) {
-                    if (scoresPtr[box_idx] > score_threshold)
-                        scores_vector.push_back(std::make_pair(scoresPtr[box_idx], box_idx));
-                }
+        const int &num_batches = static_cast<int>(scores_dims[0]);
+        const int &num_classes = static_cast<int>(scores_dims[1]);
+
+        std::vector<filteredBoxes> filtBoxes(max_output_boxes_per_class * num_batches * num_classes);
+
+        auto coeff = [&](float iou) {
+            const float weight = std::exp(scale * iou * iou);
+            return iou <= iou_threshold ? weight : 0.0f;
+        };
+
+        auto less = [](const boxInfo& l, const boxInfo& r) {
+            return l.score < r.score || ((l.score == r.score) && (l.idx > r.idx));
+        };
+
+        parallel_for2d(num_batches, num_classes, [&](int B, int C) {
```

Comment:
B and C are current indices. Not the start positions. There is no need to iterate by num_batches and num_classes.
Please look at usage examples in other layers.

Reply:
fixed